openstack-mixin/panels.libsonnet

```diff
@@ -0,0 +1,1004 @@
+        'IPs used',
+        targets=[t.ipsUsed],
+        description='The usage of available IP addresses broken down by subnet.',
+      )
+      + timeSeries.standardOptions.withUnit('percentunit')
+      + timeSeries.options.legend.withPlacement('right')
+      + timeSeries.standardOptions.withMax(1)
+      + timeSeries.standardOptions.withMin(0),
+
+    securityGroups:
+      commonlib.panels.generic.timeSeries.base.new(
+        'Security groups',
+        targets=[t.securityGroups],
+        description='The number of network security groups managed by Neutron.',
+      )
+      + timeSeries.standardOptions.withDecimals(0),
+
+    neutronAgents:
+      table.new('Agents')
+      + table.queryOptions.withTargets([t.neutronAgentState])
+      + table.queryOptions.withDatasource('prometheus', '${prometheus_datasource}')
+      + table.panelOptions.withDescription('Details for the agents for OpenStack Neutron.')
+      + table.standardOptions.withOverridesMixin([
+        table.fieldOverride.byName.new('Admin state')
+        + table.fieldOverride.byName.withProperty('custom.displayMode', 'color-text')
+        + table.fieldOverride.byName.withPropertiesFromOptions(
+          table.standardOptions.withMappings(
+            {
+              type: 'value',
+              options: {
+                down: {
+                  color: 'red',
+                  index: 1,
+                  text: 'Down',
+                },
+                up: {
+                  color: 'green',
+                  index: 0,
+                  text: 'Up',
+                },
+              },
+            }
+          ),
+        ),
+      ])
+      + table.queryOptions.withTransformationsMixin([
+        {
+          id: 'joinByField',
+          options: {
+            byField: 'Time',
+            mode: 'outer',
+          },
+        },
+        {
+          id: 'organize',
+          options: {
+            excludeByName: {
+              Time: true,
+              Value: true,
+              __name__: true,
+              hostname: true,
+              job: true,
+            },
+            includeByName: {},
+            indexByName: {
+              Time: 6,
+              Value: 9,
+              __name__: 7,
+              adminState: 3,
+              hostname: 5,
+              id: 4,
+              instance: 1,
+              job: 8,
+              service: 0,
+              zone: 2,
+            },
+            renameByName: {
+              Time: '',
+              adminState: 'Admin state',
+              id: 'ID',
+              instance: 'Instance',
+              service: 'Service',
+              zone: 'Zone',
+            },
+          },
+        },
+      ]),
+
+    volumes:
+      commonlib.panels.generic.timeSeries.base.new(
+        'Volumes',
+        targets=[t.volumes],
+        description='The number of volumes managed by Cinder.',
+      )
+      + timeSeries.standardOptions.withDecimals(0),
+
+    volumeStatus:
```

Comment:
Looks like there are two queries here both doing the same thing, with one being a top k?

Reply:
ahhh i missed that. Ok. Is there a reason one gets the topk function and the other is left as is?

Reply:
Yes, without the topk there are 20 total statuses, so this way we always get the error statuses, plus the 5 most frequent other statuses.